File: tools/run_shell.py
"""Shell sandbox — run commands via PowerShell."""

import subprocess
from tools import tool

MAX_RESULT_LEN = 2000

# Dangerous patterns to block
_BLOCKED = [
    "Remove-Item -Recurse -Force C:\\",
    "format ",
    "shutdown",
    "Restart-Computer",
    "Stop-Computer",
    "Remove-Item -Recurse -Force $env:",
    "Invoke-Expression",  # prevents indirect command injection
    "iex ",
]


def _truncate(s: str) -> str:
    return s if len(s) <= MAX_RESULT_LEN else s[:MAX_RESULT_LEN] + "... (truncated)"


@tool(
    name="run_shell",
    description=(
        "Run shell commands via PowerShell. Use for file operations, system info, "
        "process management, or any CLI task. Timeout: 10 seconds.\n"
        "Use PowerShell native syntax (Get-ChildItem, not ls). "
        "Keep commands simple — max 2 pipe stages. "
        "No interactive commands (vim, nano, pause). "
        "No destructive operations on system files."
    ),
    parameters={
        "type": "object",
        "properties": {
            "command": {
                "type": "string",
                "description": (
                    "PowerShell command. Native syntax, simple and focused.\n"
                    "GOOD: Get-ChildItem, Get-Content file.txt, Get-Process\n"
                    "BAD:  ls | grep | awk | sort | head (bash-style)\n"
                    "BAD:  complex pipelines with 3+ stages"
                ),
            }
        },
        "required": ["command"],
    },
)
def run_shell(command: str) -> str:
    # Basic safety check
    cmd_lower = command.lower()
    for blocked in _BLOCKED:
        if blocked.lower() in cmd_lower:
            return f"Error: blocked dangerous command pattern: {blocked}"

    try:
        proc = subprocess.run(
            ["pwsh", "-NoProfile", "-Command", command],
            capture_output=True,
            text=True,
            timeout=10,
            encoding="utf-8",
            errors="replace",
        )
        stdout = proc.stdout.strip()
        stderr = proc.stderr.strip()

        if proc.returncode != 0 and stderr:
            return f"Error: {stderr}"
        if stdout:
            return _truncate(stdout)
        if stderr:
            return _truncate(stderr)
        return "(no output)"
    except subprocess.TimeoutExpired:
        return "timeout (10s)"
    except FileNotFoundError:
        return "Error: PowerShell not found"

File: tools/run_shell.ts
/** Shell sandbox — run commands via PowerShell. */

import { z } from "zod";
import { registerTool } from "./lib/registry.js";
import { spawnProcess, truncate } from "./lib/env_helpers.js";
import { SandboxError } from "./lib/errors.js";

const TIMEOUT = 10_000;

// Dangerous patterns to block (same as Python version)
const BLOCKED = [
  "Remove-Item -Recurse -Force C:\\",
  "format ",
  "shutdown",
  "Restart-Computer",
  "Stop-Computer",
  "Remove-Item -Recurse -Force $env:",
  "Invoke-Expression", // prevents indirect command injection
  "iex ",
];

async function runShell(command: string): Promise<string> {
  // Basic safety check
  const cmdLower = command.toLowerCase();
  for (const blocked of BLOCKED) {
    if (cmdLower.includes(blocked.toLowerCase())) {
      return `Error: blocked dangerous command pattern: ${blocked}`;
    }
  }

  // Force UTF-8 output encoding for pwsh to handle non-ASCII correctly
  const wrapped = "[Console]::OutputEncoding = [Text.Encoding]::UTF8; " + command;

  try {
    const result = await spawnProcess("pwsh", ["-NoProfile", "-Command", wrapped], TIMEOUT);
    if (result.timedOut) throw SandboxError.timeout();

    const stdout = result.stdout.trim();
    const stderr = result.stderr.trim();

    if (result.exitCode !== 0 && stderr) throw SandboxError.executionError(`Error: ${stderr}`);
    if (stdout) return truncate(stdout);
    if (stderr) return truncate(stderr);
    return "(no output)";
  } catch (e) {
    if (e instanceof SandboxError) throw e;
    throw SandboxError.notFound("PowerShell");
  }
}

const DESCRIPTION =
  "Run shell commands via PowerShell. Use for file operations, system info, " +
  "process management, or any CLI task. Timeout: 10 seconds.\n" +
  "Use PowerShell native syntax (Get-ChildItem, not ls). " +
  "Keep commands simple — max 2 pipe stages. " +
  "No interactive commands (vim, nano, pause). " +
  "No destructive operations on system files.";

const COMMAND_DESC =
  "PowerShell command. Native syntax, simple and focused.\n" +
  "GOOD: Get-ChildItem, Get-Content file.txt, Get-Process\n" +
  "BAD:  ls | grep | awk | sort | head (bash-style)\n" +
  "BAD:  complex pipelines with 3+ stages";

registerTool({
  name: "run_shell",
  description: DESCRIPTION,
  parameters: {
    command: z.string().describe(COMMAND_DESC),
  },
  handler: async (args) => {
    try {
      return await runShell(args.command as string);
    } catch (e) {
      if (e instanceof SandboxError) return e.message;
      throw e;
    }
  },
});
